### DISCUSSION TITLE
How should I define process.env['NODE_CONFIG_DIR'] ?

### DISCUSSION THREAD

**Discussion by ljlm0402** [swc-project/swc#2790] [Q&A] [answered]
Thank you so much for the good open source.

This is the current folder structure, and the environment variable is controlled through NODE_ENV as a json file in the configs folder.
```sh
├── /src
│   ├── /configs
│   │   ├── development.json
│   │   ├── production.json
│   │   └── test.json
│   │
│   ├── /controllers
│   │   ├── auth.controller.ts
│   │   ├── index.controller.ts
│   │   └── users.controller.ts
│   │
│   ├── /dtos
│   │   └── users.dto.ts
│   │
│   ├── /exceptions
│   │   └── HttpException.ts
│   │
│   ├── /http
│   │   ├── auth.http
│   │   └── users.http
│   │
│   ├── /interfaces
│   │   ├── auth.interface.ts
│   │   ├── routes.interface.ts
│   │   └── users.interface.ts
│   │
│   ├── /middlewares
│   │   ├── auth.middleware.ts
│   │   ├── error.middleware.ts
│   │   └── validation.middleware.ts
│   │
│   ├── /models
│   │   └── users.model.ts
│   │
│   ├── /routes
│   │   ├── auth.route.ts
│   │   ├── index.route.ts
│   │   └── users.route.ts
│   │
│   ├── /services
│   │   ├── auth.service.ts
│   │   └── users.service.ts
│   │
│   ├── /tests
│   │   ├── auth.test.ts
│   │   ├── index.test.ts
│   │   └── users.test.ts
│   │
│   ├── /utils
│   │   ├── logger.ts
│   │   ├── util.ts
│   │   └── vaildateEnv.ts
│   │
│   ├── app.ts
│   └── server.ts
```

This is the server.js output through the existing `tsc && tsc-alias` build. As shown in the image below, process.env['NODE_CONFIG_DIR'] is defined at the top.

![image](https://user-images.githubusercontent.com/42952358/142342927-e3f07158-b4cb-4852-acf9-4d7423840b2a.png)

---

However, the swc build output is defined after config, as in the image below, and things are not recognized.

![image](https://user-images.githubusercontent.com/42952358/142343161-933b2c5d-ca0a-4448-a37a-619d0b52f238.png)

my `swcrc` file and command `swc src -d dist --source-maps --copy-files`

```sh
{
    "jsc": {
      "parser": {
        "syntax": "typescript",
        "tsx": false,
        "dynamicImport": true,
        "decorators": true
      },
      "target": "es2017",
      "externalHelpers": false,
      "keepClassNames": true,
        "loose": false,
        "minify": {
          "compress": false,
          "mangle": false
        },
      "baseUrl": "src",
      "paths": {
          "@/*": ["*"],
          "@configs/*": ["configs/*"],
          "@controllers/*": ["controllers/*"],
          "@dtos/*": ["dtos/*"],
          "@exceptions/*": ["exceptions/*"],
          "@interfaces/*": ["interfaces/*"],
          "@middlewares/*": ["middlewares/*"],
          "@models/*": ["models/*"],
          "@routes/*": ["routes/*"],
          "@services/*": ["services/*"],
          "@utils/*": ["utils/*"]
        }
    },
    "module": {
      "type": "commonjs"
    },
    "minify": false
  }
```

I don't know what to do ㅠ_ㅠ. Any help would be appreciated.

**Comment by ljlm0402** [comment]
+ This is an error message that occurs when running server.js.

![image](https://user-images.githubusercontent.com/42952358/142343684-9c10e3f3-de73-4a5a-998d-72dff59c9b7d.png)

`configs/*.json` file info
```
{
    "log": {
        "format": "dev",
        "dir": "../logs"
    }
}
```

**Comment by kwonoj** [comment, marked as answer]
Would you share how original import is composed? This sounds like same as https://github.com/swc-project/swc/discussions/2729 but cannot confirm without seeing original code.

**Reply by ljlm0402** [reply]
this `server.ts` file 

![image](https://user-images.githubusercontent.com/42952358/142365786-00fb981a-0063-4538-8e31-e9ea4d348af8.png)

If you want to perform additional tests, it may be helpful to install and proceed with the 
http://npm.im/typescript-express-starter open source `default` template.

**Reply by kwonoj** [reply]
Yes, this is #2729. As I shared in 2729, swc models babel which does hosting for import correctly.

The reason it works on tsc is due to https://github.com/microsoft/TypeScript/issues/16166 which tsc doesn't hoist import correctly.

**Reply by ljlm0402** [reply]
@kwonoj 

Thanks for letting us know about the cause. First of all, let's come up with another way to proceed. I'm just wondering if there is any good way to do it.

**Comment by kdy1** [comment]
Seems like a regression of documentation...
